src/index.js

```diff
@@ -0,0 +1,196 @@
+    if (t.isJSXMemberExpression(node)) {
+      return t.inherits(
+        t.memberExpression(jsxToJS(node.object), jsxToJS(node.property)),
+        node
+      );
+    }
+    return t.inherits(t.identifier(node.name), node);
+  }
+
+  /**
+   * @param {import("@babel/types").JSXIdentifier} node
+   * @returns {import("@babel/types").StringLiteral}
+   */
+  function jsxToString(node) {
+    return t.inherits(t.stringLiteral(node.name), node);
+  }
+
+  function transformAttributesList(path) {
+    /** @type {import("@babel/types").JSXOpeningElement} */
+    const node = path.node;
+
+    if (node.attributes.length === 0) return t.nullLiteral();
+
+    let hasSpread, hasDynamic, hasStatic;
+    for (const attr of node.attributes) {
+      if (t.isJSXSpreadAttribute(attr)) {
+        hasSpread = true;
+        break;
+      } else if (
+        t.isJSXAttribute(attr) &&
+        t.isJSXExpressionContainer(attr.value)
+      ) {
+        hasDynamic = true;
+      } else {
+        hasStatic = true;
+      }
+    }
+    const type =
+      hasSpread || !hasStatic
+        ? "RUNTIME_TYPE"
+        : hasDynamic
+        ? "MIXED_TYPE"
+        : "STATIC_TYPE";
+
+    return template.expression.ast`
+      ESXToken.create(
+        ESXToken.${t.identifier(type)},
+        ${t.arrayExpression(path.get("attributes").map(transformAttribute))}
+      )
+    `;
+  }
+
+  function transformAttribute(path) {
+    /** @type {import("@babel/types").JSXAttribute | import("@babel/types").JSXSpreadAttribute} */
+    const node = path.node;
+
+    let type, name, value;
+    if (t.isJSXSpreadAttribute(node)) {
+      type = "RUNTIME_TYPE";
+      name = t.stringLiteral("TODO: spread");
+      value = node.argument;
+    } else if (t.isJSXExpressionContainer(node.value)) {
+      type = "RUNTIME_TYPE";
+      name = jsxToString(node.name);
+      value = node.value.expression;
+    } else if (t.isJSXElement(node.value) || t.isJSXFragment(node.value)) {
+      throw path
+        .get("value")
+        .buildCodeFrameError(
+          "JSX elements are not supported as static attributes. Please wrap it in { }."
+        );
+    } else if (node.value) {
+      type = "STATIC_TYPE";
+      name = jsxToString(node.name);
+      value = node.value;
+    } else {
+      type = "STATIC_TYPE";
+      name = jsxToString(node.name);
+      value = t.booleanLiteral(true);
+    }
+
+    return t.inherits(
+      template.expression.ast`
+        ESXToken.property(ESXToken.${t.identifier(type)}, ${name}, ${value})
+      `,
+      node
+    );
+  }
+
+  function transformChild(path) {
+    /** @type {import("@babel/types").JSXElement["children"][number]} */
+    const node = path.node;
+
+    let type, value;
+    if (t.isJSXExpressionContainer(node)) {
+      type = "RUNTIME_TYPE";
+      value = node.expression;
+    } else if (t.isJSXSpreadChild(node)) {
+      // <div>{...foo}</div>
+      throw path.buildCodeFrameError("Spread children are not supported.");
```

Comment:
Do you want to support this? (the default Babel plugin for JSX doesn't)
What should the output look like?

Reply:
Spread children are not supported, an array would be jus fine, no need to spread